### ISSUE TITLE
Questions about Oauth

### ISSUE THREAD

**Issue by MrAmericanMike** [spotify/web-api#1028]
Sorry if this is not the place to ask for some guidance, but the support page points me here.

I'm developing an application on NodeJS and Electron that will become a desktop application. I'm having issues figuring out what the best way to do the Oauth proccess will be. At first I was using an electron-oauth library, but apparently this is not the way to go, as it uses the client-secret and apparently it's not a good idea to have that on desktop apps.

What route should I take here? What's the best way to authorize my application?

Thanks for the help.

**Comment by 0xNF** [comment]
You'll likely want to look into using the node library [Passport](http://passportjs.org/), with the [Spotify](https://github.com/JMPerez/passport-spotify) plugin for handling OAuth on Node.  

Second, if you don't intend to have a server, that is, if the entirety of your app will live in the Electron instance, then Client Credentials flow is good enough - it's certainly a supported flow, it's just less secure (but also less complicated) than the Authorization Code flow.

**Comment by MrAmericanMike** [comment]
> 
> 
> You'll likely want to look into using the node library [Passport](http://passportjs.org/), with the [Spotify](https://github.com/JMPerez/passport-spotify) plugin for handling OAuth on Node.
> 
> Second, if you don't intend to have a server, that is, if the entirety of your app will live in the Electron instance, then Client Credentials flow is good enough - it's certainly a supported flow, it's just less secure (but also less complicated) than the Authorization Code flow.

The problem that I see is that doing only ClientID I can get Access Token, but not a Refresh Token. So the user would have to relog every 30 minutes.

Am I missing something, or I'm looking at the wrong method?

Thanks a lot. Sorry I'm new to all this Oauth stuff.

**Comment by 0xNF** [comment]
Well, if you want you can absolutely do the Authorization Code flow on the client. It's "not recommended", but it's totally doable.

Just handle all the steps in your electron app and store the Refresh Token someplace safe.

**Comment by MrAmericanMike** [comment]
> 
> 
> Well, if you want you can absolutely do the Authorization Code flow on the client. It's "not recommended", but it's totally doable.
> 
> Just handle all the steps in your electron app and store the Refresh Token someplace safe.

Problem is that the secret key will be distributed with the final app that way.

I'm checking if I can get the regular token every hour by just requesting it away, apparently if you select "Remember me" (I forgot how the mark is called) when you get redirected to the Auth page, you get redirected back to the callback with a token. Just need to check if I can do that on an non intrusive way.

**Comment by 0xNF** [comment]
Yes, the client secret is definitely distributed with the final app, which is why it's not recommended. 

You could stand up a basic server with something small like a sqlite database that handles these requests, which would leave the secret on your server instead. It adds some complexity to whole process though for sure, and maybe you can get by with the remember me flag instead.

**Comment by MrAmericanMike** [comment]
So... Talking with a friend we were wondering. What's the problem about the secret key been visible to others, if even when using only the ClientID others can get their hand on it and authorize users against Spotify as if they were "My app" or for whatever reasons.
What makes "distributing" the ClientID not a problem as "distributing" the Client Secret?

At the end, even with only the ClientID anyone can authorize against Spotify as if they were "me"

**Comment by 0xNF** [comment]
It opens the door for an attacker to impersonate your application and gain access to any permissions a user has authorized your application for. With a Spotify account, that could be anywhere from deleting their library and playlists, to seeing email and birthdate information. 

This would require the attacker to also grab the access token the next time they log in, and the users would also have to log in to the fraudulent application as opposed to your own for this to happen.

There are a number of things to have to happen together before the risks of an exposed client secret manifest, but you're certainly making it easier for an attacker by revealing your key.  

It's really up to you to determine how much of a risk this, which depends on what kind of permissions you've asked your users to grant, but you're not doing your users any favors by cutting corners here.

**Comment by MrAmericanMike** [comment]
> 
> 
> It opens the door for an attacker to impersonate your application and gain access to any permissions a user has authorized your application for. With a Spotify account, that could be anywhere from deleting their library and playlists, to seeing email and birthdate information.
> 
> This would require the attacker to also grab the access token the next time they log in, and the users would also have to log in to the fraudulent application as opposed to your own for this to happen.
> 
> There are a number of things to have to happen together before the risks of an exposed client secret manifest, but you're certainly making it easier for an attacker by revealing your key.
> 
> It's really up to you to determine how much of a risk this, which depends on what kind of permissions you've asked your users to grant, but you're not doing your users any favors by cutting corners here.
> 
> edit:
> Of course, an attacker will also be doing all of this on your ID, meaning Spotify will come after you for any abuses.

Well but... unless an attacker gets a user to "relog" using my clientID and clientSecret they can't do anything with just those 2 right? Also, and going to the point about the difference between just exposing clientID or clientID + clientSecret
I see no difference, both method can make it so anyone can use them to make a user log into X app thinking they are allowing my app.

So yes, not trying to cut corners here, the only alternative is to have a server where I can redirect the users so I get the tokens from my own server, right? (And even then... anyone could send requests to my server and get authorization codes?) I'm confuse.

**Comment by 0xNF** [comment]
It's a minor difference to be sure, but a malicious actor can't do anything with a client id. If some hypothetical attacker were to stand up their own application styled to look like yours, a user would have to authorize all the permissions, just like they would when logging into your application the first time. 

Having your client secret, they can skip that additional authorization step and have what, to the user, looks like a transparent pass-through. Just like how logging into your application a second time doesn't prompt a re-authorization of permissions, that step would be skipped in this scenario too.

You're right that the maximum damage of an exposed client secret is minimal. Because every API Spotify offers must be done with a user token the ability to wreak havoc is diminished.

The other thing to consider is that if your application _does_ get compromised and you're distributing the client secret with your application, then you'd have to distribute a whole new updated client to remedy the situation, and users would remain vulnerable until they received the update. In contrast, a key stored server-side asks nothing of the user when regenerated.

**Comment by antice-eton** [comment]
Does Spotify support PKCE?
